presto-jdbc/src/test/java/io/prestosql/jdbc/TestPrestoDriverExternalAuth.java

```diff
@@ -0,0 +1,482 @@
+    public void testAuthenticationAfterInitialTokenHasBeenInvalidated()
+            throws Exception
+    {
+        try (RedirectHandlerFixture ignore = withHandler(HTTPGetOnlyRedirectHandler::new);
+                Connection connection = createConnection(ImmutableMap.of("externalAuthentication", "true"));
+                Statement statement = connection.createStatement()) {
+            statement.execute("SELECT 123");
+
+            invalidateAllTokens();
+            assertThat(countIssuedTokens()).isEqualTo(0);
+
+            assertThat(statement.execute("SELECT 123")).isTrue();
+        }
+    }
+
+    private Connection createConnection(Map<String, String> additionalProperties)
+            throws SQLException
+    {
+        String url = format("jdbc:presto://localhost:%s", server.getHttpsAddress().getPort());
+        Properties properties = new Properties();
+        properties.setProperty("user", "test");
+        properties.setProperty("SSL", "true");
+        properties.setProperty("SSLTrustStorePath", getResource("localhost.truststore").getPath());
+        properties.setProperty("SSLTrustStorePassword", "changeit");
+        additionalProperties.forEach(properties::setProperty);
+        return DriverManager.getConnection(url, properties);
+    }
+
+    static class DummyExternalAuthModule
+            extends AbstractConfigurationAwareModule
+    {
+        private final IntSupplier port;
+        private TokenPollingErrorFixture errorFixture;
+
+        DummyExternalAuthModule(IntSupplier port, TokenPollingErrorFixture errorFixture)
+        {
+            this.port = requireNonNull(port, "port is null");
+            this.errorFixture = requireNonNull(errorFixture, "errorFixture is null");
+        }
+
+        @Override
+        protected void setup(Binder binder)
+        {
+            install(authenticatorModule("dummy-external", DummyAuthenticator.class, subBinder -> {
+                subBinder.bind(Authentications.class).in(SINGLETON);
+                subBinder.bind(IntSupplier.class).toInstance(port);
+                jaxrsBinder(subBinder).bind(DummyExternalAuthResources.class);
+            }));
+        }
+    }
+
+    static class Authentications
+    {
+        private final ConcurrentHashMap<String, String> logginSessions = new ConcurrentHashMap<>();
+        private final Set<String> validTokens = ConcurrentHashMap.newKeySet();
+
+        String startAuthentication()
+        {
+            String sessionId = UUID.randomUUID().toString();
+            logginSessions.put(sessionId, "");
+            return sessionId;
+        }
+
+        public void logIn(String sessionId)
+        {
+            String token = sessionId + "_token";
+            validTokens.add(token);
+            logginSessions.put(sessionId, token);
+        }
+
+        public Optional<String> getToken(String sessionId)
+                throws IllegalArgumentException
+        {
+            if (logginSessions.containsKey(sessionId)) {
```

Comment:
Using contains+get is not safe for a concurrent map. Fetch the value and check for null